src/engine/enginemaster.cpp

```diff
@@ -443,48 +530,212 @@ void EngineMaster::process(const int iBufferSize) {
     }
 
     if (masterEnabled) {
-        // Mix the three channels together. We already mixed the busses together
-        // with the channel gains and overall master gain.
-        if (!m_pMasterTalkoverMix->toBool()) {
-            // Add Talkover to Master output
-            SampleUtil::copy4WithGain(m_pMaster,
-                    m_pOutputBusBuffers[EngineChannel::LEFT], 1.0,
-                    m_pOutputBusBuffers[EngineChannel::CENTER], 1.0,
-                    m_pOutputBusBuffers[EngineChannel::RIGHT], 1.0,
+        // Mix the crossfader orientation buffers together into the master mix
+        SampleUtil::copy3WithGain(m_pMaster,
+            m_pOutputBusBuffers[EngineChannel::LEFT], 1.0,
+            m_pOutputBusBuffers[EngineChannel::CENTER], 1.0,
+            m_pOutputBusBuffers[EngineChannel::RIGHT], 1.0,
+            iBufferSize);
+
+        // Process master, booth, and record/broadcast buffers according to the
+        // TalkoverMixMode configured in DlgPrefSound
+
+        // Mixxx gets input signals delayed by the input latency,
+        // so to preserve the relative timing of the input with Mixxx's outputs
+        // (headphones, master, booth, and record/broadcast), delay the outputs
+        // by the input latency before mixing the inputs into them.
+
+        // This way, if the microphone user is on beat with the output they hear from
+        // the speakers/headphones, the input will remain on beat at the expense of
+        // adding some latency to Mixxx's output (no latency will be added to the
+        // microphone signal though). If no microphone inputs are configured, avoid adding
+        // this extra latency.
+
+        // If the input signal is being directly monitored (mixed with the master output
+        // in hardware), do not delay the audible outputs. However, delay the
+        // record/broadcast signal to keep it on beat and consistent with what
+        // the user hears from the speakers/headphones.
+        if (configuredTalkoverMixMode == TalkoverMixMode::MASTER) {
+            // Apply input latency compensation to master mix
+            // Do not add unnecessary latency if no microphones are configured
+            if (m_pNumMicsConfigured->get() > 0) {
+                m_pInputLatencyCompensationDelay->process(m_pMaster, iBufferSize);
+            }
+
+            // Copy master mix to booth output with booth gain before mixing
+            // talkover with master mix
+            if (boothEnabled) {
+                CSAMPLE boothGain = m_pBoothGain->get();
+                if (m_bRampingGain) {
+                    SampleUtil::copy1WithRampingGain(m_pBooth, m_pMaster,
```

Comment:
>>  That is the desired effect.

> Huh? Why would that ever be desired?

It is desired to have two time domains, the voice / monitor time domain with a minimum latency and the master time domain with any latency but the voice synced as haired in the first time domain.   

>> For this setup the user must be isolated from the master sound to not hear the double beat effect.

> Isolating the DJ booth from the master sound does not happen and does not make sense as that would also disconnect the performer from the audience.

It actually happens as soon you use close headphones, or play the booth / headphones significant louder than the sound that reaches you form the main PA system. For the audiance feedback you can just take off the headphones or lower the booth volume. 

>> If this cannot be achieved, the latency compensation does not work.

> I don't think you're understanding what latency compensation is for. Have you tested this yet?

It really looks like, we have different ideas from latency compensation. I was not able to test is up to now, because the comensation delay was always added on to the headphones and booth. What is your understanding of latency compensation? What is the problem you suffer without it and how does you implementation solves the issue? 

>> In this case I would recommend to add a Booth and Headphone delay to compensate the distance to the main PA speakers.

> If Mixxx does not perform the same latency compensation on all outputs, that creates an artificial situation where the user must understand the calculation Mixxx does on the Measured Round Trip Latency, perform that calculation themselves, and configure the headphone and booth delays to match that just as a baseline before using the output delays for their intended purpose (compensating for speaker placement). The PFL mix must be latency compensated or the same timing issue that currently occurs on the master output without this branch would occur for mics routed to PFL.

I do not understand that. In my theory, a delayed MonitorMix in headpones destroys the compensation effect, because it is from the master time domain, but the headphones belongs to the voice time domain. For my use-case, you can just measure the roundtrip latency and use that for compensation. Any additional unknown latency of any reason it included and though compensated.